File: src/primitives/address/address_test.zig
const std = @import("std");
const testing = std.testing;
const Address = @import("address.zig");

// Test address validation
test "is_valid_address" {
    // Valid addresses (lowercase)
    try testing.expect(Address.is_valid("0x0000000000000000000000000000000000000000"));
    try testing.expect(Address.is_valid("0xa0cf798816d4b9b9866b5330eea46a18382f251e"));
    try testing.expect(Address.is_valid("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"));
    
    // Valid addresses (checksummed)
    try testing.expect(Address.is_valid("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"));
    try testing.expect(Address.is_valid("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"));
    
    // Invalid addresses
    try testing.expect(!Address.is_valid("x"));
    try testing.expect(!Address.is_valid("0x"));
    try testing.expect(!Address.is_valid("0xa"));
    try testing.expect(!Address.is_valid("0xa5cc3c03994db5b0d9a5eEdD10Cabab0813678az")); // Invalid character
    try testing.expect(!Address.is_valid("0xa5cc3c03994db5b0d9a5eEdD10Cabab0813678aff")); // Too long
    try testing.expect(!Address.is_valid("a5cc3c03994db5b0d9a5eEdD10Cabab0813678ac")); // Missing 0x
}

// Test address checksum validation
test "validate_checksum" {
    // Valid checksummed addresses
    try testing.expect(try Address.validate_checksum("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"));
    try testing.expect(try Address.validate_checksum("0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266"));
    try testing.expect(try Address.validate_checksum("0x70997970C51812dc3A010C7d01b50e0d17dc79C8"));
    try testing.expect(try Address.validate_checksum("0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC"));
    try testing.expect(try Address.validate_checksum("0x90F79bf6EB2c4f870365E785982E1f101E93b906"));
    try testing.expect(try Address.validate_checksum("0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65"));
    
    // Invalid checksum
    try testing.expect(!try Address.validate_checksum("0xa5cc3c03994db5b0d9a5eEdD10Cabab0813678ac"));
    
    // All lowercase is valid (no checksum)
    try testing.expect(try Address.validate_checksum("0xa0cf798816d4b9b9866b5330eea46a18382f251e"));
    
    // All uppercase is valid (no checksum)
    try testing.expect(try Address.validate_checksum("0xA0CF798816D4B9B9866B5330EEA46A18382F251E"));
}

// Test address checksum computation
test "compute_checksum" {
    const allocator = testing.allocator;
    
    // Test vectors from EIP-55
    const test_cases = .{
        .{ "0xa0cf798816d4b9b9866b5330eea46a18382f251e", "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e" },
        .{ "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266", "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266" },
        .{ "0x70997970c51812dc3a010c7d01b50e0d17dc79c8", "0x70997970C51812dc3A010C7d01b50e0d17dc79C8" },
        .{ "0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc", "0x3C44CdDdB6a900fa2b585dd299e03d12FA4293BC" },
        .{ "0x90f79bf6eb2c4f870365e785982e1f101e93b906", "0x90F79bf6EB2c4f870365E785982E1f101E93b906" },
        .{ "0x15d34aaf54267db7d7c367839aaf71a00a2c6a65", "0x15d34AAf54267DB7D7c367839AAf71A00a2C6A65" },
        .{ "0xa5cc3c03994db5b0d9a5eedd10cabab0813678ac", "0xa5cc3c03994DB5b0d9A5eEdD10CabaB0813678AC" },
    };
    
    inline for (test_cases) |test_case| {
        const checksummed = try Address.to_checksum(allocator, test_case[0]);
        defer allocator.free(checksummed);
        try testing.expectEqualStrings(test_case[1], checksummed);
    }
}

// Test address parsing from hex
test "from_hex" {
    // Valid address
    const addr1 = try Address.from_hex("0xa0cf798816d4b9b9866b5330eea46a18382f251e");
    const expected1 = [20]u8{
        0xa0, 0xcf, 0x79, 0x88, 0x16, 0xd4, 0xb9, 0xb9,
        0x86, 0x6b, 0x53, 0x30, 0xee, 0xa4, 0x6a, 0x18,
        0x38, 0x2f, 0x25, 0x1e,
    };
    try testing.expectEqual(expected1, addr1);
    
    // Address with mixed case (should work)
    const addr2 = try Address.from_hex("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e");
    try testing.expectEqual(expected1, addr2);
    
    // Invalid length
    const result1 = Address.from_hex("0xa0cf798816d4b9b9866b5330eea46a18382f251");
    try testing.expectError(Address.AddressError.InvalidLength, result1);
    
    // Invalid prefix
    const result2 = Address.from_hex("a0cf798816d4b9b9866b5330eea46a18382f251e");
    try testing.expectError(Address.AddressError.InvalidFormat, result2);
}

// Test address to hex conversion
test "to_hex" {
    const allocator = testing.allocator;
    
    const addr = [20]u8{
        0xa0, 0xcf, 0x79, 0x88, 0x16, 0xd4, 0xb9, 0xb9,
        0x86, 0x6b, 0x53, 0x30, 0xee, 0xa4, 0x6a, 0x18,
        0x38, 0x2f, 0x25, 0x1e,
    };
    
    // Lowercase
    const lower = try Address.to_hex(allocator, addr);
    defer allocator.free(lower);
    try testing.expectEqualStrings("0xa0cf798816d4b9b9866b5330eea46a18382f251e", lower);
    
    // Uppercase
    const upper = try Address.to_hex_upper(allocator, addr);
    defer allocator.free(upper);
    try testing.expectEqualStrings("0xA0CF798816D4B9B9866B5330EEA46A18382F251E", upper);
    
    // Checksummed
    const checksummed = try Address.to_checksum_from_bytes(allocator, addr);
    defer allocator.free(checksummed);
    try testing.expectEqualStrings("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e", checksummed);
}

// Test address comparison
test "address equality" {
    const addr1 = try Address.from_hex("0xa0cf798816d4b9b9866b5330eea46a18382f251e");
    const addr2 = try Address.from_hex("0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"); // Same but different case
    const addr3 = try Address.from_hex("0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266");
    
    try testing.expect(Address.equal(addr1, addr2));
    try testing.expect(!Address.equal(addr1, addr3));
    try testing.expect(Address.equal(addr1, addr1));
}

// Test zero address
test "is_zero_address" {
    const zero = try Address.from_hex("0x0000000000000000000000000000000000000000");
    try testing.expect(Address.is_zero(zero));
    
    const non_zero = try Address.from_hex("0x0000000000000000000000000000000000000001");
    try testing.expect(!Address.is_zero(non_zero));
}

// Test address constants
test "address constants" {
    try testing.expectEqual([20]u8{0} ** 20, Address.ZERO);
    try testing.expect(Address.is_zero(Address.ZERO));
}

// Test create2 address computation
test "compute_create2_address" {
    const allocator = testing.allocator;
    
    // Test vector from EIP-1014
    const from = try Address.from_hex("0x0000000000000000000000000000000000000000");
    const salt = [32]u8{0} ** 32;
    const init_code_hash = [32]u8{0} ** 32;
    
    const result = try Address.compute_create2_address(allocator, from, salt, init_code_hash);
    
    // Should produce a deterministic address
    const result2 = try Address.compute_create2_address(allocator, from, salt, init_code_hash);
    try testing.expectEqual(result, result2);
}

// Test address from public key
test "from_public_key" {
    // Test vector: public key to address conversion
    // This is a known test vector where we know the public key and expected address
    const public_key = [64]u8{
        // X coordinate
        0x9d, 0x61, 0xb1, 0x9d, 0xef, 0xfd, 0x5a, 0x60,
        0xba, 0x84, 0x4a, 0xf4, 0x92, 0xec, 0x2c, 0xc4,
        0x44, 0x49, 0xc5, 0x69, 0x7b, 0x32, 0x69, 0x19,
        0x70, 0x3b, 0xac, 0x03, 0x1c, 0xae, 0x7f, 0x60,
        // Y coordinate  
        0x47, 0x77, 0xef, 0x49, 0xde, 0xad, 0xbe, 0xef,
        0x8e, 0xf8, 0x6e, 0x3e, 0x3f, 0xef, 0x3a, 0x6f,
        0x26, 0x7e, 0x35, 0x65, 0x8a, 0x9c, 0x1a, 0x17,
        0x27, 0x44, 0xae, 0x81, 0x5f, 0x89, 0xf7, 0xef,
    };
    
    const addr = Address.from_public_key(public_key);
    
    // Address should be last 20 bytes of keccak256(public_key)
    try testing.expect(!Address.is_zero(addr));
}

// Test address string equality (case insensitive)
test "address string equality" {
    const allocator = testing.allocator;
    
    try testing.expect(try Address.equal_strings(allocator, 
        "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
        "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e"
    ));
    
    try testing.expect(!try Address.equal_strings(allocator,
        "0xa0cf798816d4b9b9866b5330eea46a18382f251e",
        "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266"
    ));
}

// Test invalid address formats
test "invalid address formats" {
    const test_cases = [_][]const u8{
        "", // Empty
        "0x", // Only prefix
        "0xG0cf798816d4b9b9866b5330eea46a18382f251e", // Invalid character G
        "00a0cf798816d4b9b9866b5330eea46a18382f251e", // Missing 0x prefix
        "0xa0cf798816d4b9b9866b5330eea46a18382f251", // Too short (39 chars)
        "0xa0cf798816d4b9b9866b5330eea46a18382f251ee", // Too long (41 chars)
        "0xa0cf798816d4b9b9866b5330eea46a18382f251 ", // Space at end
        " 0xa0cf798816d4b9b9866b5330eea46a18382f251e", // Space at start
    };
    
    for (test_cases) |test_case| {
        const result = Address.from_hex(test_case);
        try testing.expectError(Address.AddressError.InvalidFormat, result);
    }
}

// Test batch address operations
test "validate multiple addresses" {
    const allocator = testing.allocator;
    
    const addresses = [_][]const u8{
        "0xA0Cf798816D4b9b9866b5330EEa46a18382f251e",
        "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266",
        "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
    };
    
    // All should be valid
    for (addresses) |addr| {
        try testing.expect(Address.is_valid(addr));
        try testing.expect(try Address.validate_checksum(addr));
    }
}

// Test known Ethereum addresses
test "well known addresses" {
    // Test some well-known Ethereum addresses
    const null_address = "0x0000000000000000000000000000000000000000";
    try testing.expect(Address.is_valid(null_address));
    
    const dead_address = "0x000000000000000000000000000000000000dEaD";
    try testing.expect(Address.is_valid(dead_address));
    try testing.expect(try Address.validate_checksum(dead_address));
}

File: src/primitives/Address/Address_test.go
package address

import (
	"math/big"
	"testing"
)

func TestFromHex(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		wantErr bool
	}{
		{
			name:    "valid lowercase hex",
			input:   "0x742d35cc6634c0532925a3b844bc9e7595f251e3",
			wantErr: false,
		},
		{
			name:    "valid mixed case hex",
			input:   "0x742d35Cc6634C0532925a3b844Bc9e7595f251e3",
			wantErr: false,
		},
		{
			name:    "zero address",
			input:   "0x0000000000000000000000000000000000000000",
			wantErr: false,
		},
		{
			name:    "max address",
			input:   "0xffffffffffffffffffffffffffffffffffffffff",
			wantErr: false,
		},
		{
			name:    "missing 0x prefix",
			input:   "742d35cc6634c0532925a3b844bc9e7595f251e3",
			wantErr: true,
		},
		{
			name:    "invalid length too short",
			input:   "0x742d35cc",
			wantErr: true,
		},
		{
			name:    "invalid length too long",
			input:   "0x742d35cc6634c0532925a3b844bc9e7595f251e3ff",
			wantErr: true,
		},
		{
			name:    "invalid hex characters",
			input:   "0x742d35cc6634c0532925a3b844bc9e7595f251eZ",
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := FromHex(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("FromHex() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr && len(addr) != Size {
				t.Errorf("FromHex() length = %d, want %d", len(addr), Size)
			}
		})
	}
}

func TestFromBytes(t *testing.T) {
	tests := []struct {
		name    string
		input   []byte
		wantErr bool
	}{
		{
			name:    "valid 20 bytes",
			input:   make([]byte, 20),
			wantErr: false,
		},
		{
			name:    "invalid length 19 bytes",
			input:   make([]byte, 19),
			wantErr: true,
		},
		{
			name:    "invalid length 21 bytes",
			input:   make([]byte, 21),
			wantErr: true,
		},
		{
			name:    "nil bytes",
			input:   nil,
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := FromBytes(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("FromBytes() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr && len(addr) != Size {
				t.Errorf("FromBytes() length = %d, want %d", len(addr), Size)
			}
		})
	}
}

func TestFromNumber(t *testing.T) {
	tests := []struct {
		name    string
		input   *big.Int
		wantErr bool
	}{
		{
			name:    "zero",
			input:   big.NewInt(0),
			wantErr: false,
		},
		{
			name:    "small positive number",
			input:   big.NewInt(12345),
			wantErr: false,
		},
		{
			name:    "negative number",
			input:   big.NewInt(-1),
			wantErr: true,
		},
		{
			name:    "large number",
			input:   new(big.Int).SetBytes([]byte{0x74, 0x2d, 0x35, 0xcc, 0x66, 0x34, 0xc0, 0x53, 0x29, 0x25, 0xa3, 0xb8, 0x44, 0xbc, 0x9e, 0x75, 0x95, 0xf2, 0x51, 0xe3}),
			wantErr: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			addr, err := FromNumber(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("FromNumber() error = %v, wantErr %v", err, tt.wantErr)
				return
			}
			if !tt.wantErr && len(addr) != Size {
				t.Errorf("FromNumber() length = %d, want %d", len(addr), Size)
			}
		})
	}
}

func TestFrom(t *testing.T) {
	tests := []struct {
		name    string
		input   interface{}
		wantErr bool
	}{
		{
			name:    "from string",
			input:   "0x742d35cc6634c0532925a3b844bc9e7595f251e3",
			wantErr: false,
		},
		{
			name:    "from int",
			input:   12345,
			wantErr: false,
		},
		{
			name:    "from int64",
			input:   int64(12345),
			wantErr: false,
		},
		{
			name:    "from uint64",
			input:   uint64(12345),
			wantErr: false,
		},
		{
			name:    "from big.Int",
			input:   big.NewInt(12345),
			wantErr: false,
		},
		{
			name:    "from bytes",
			input:   make([]byte, 20),
			wantErr: false,
		},
		{
			name:    "from Address",
			input:   Address{},
			wantErr: false,
		},
		{
			name:    "unsupported type",
			input:   true,
			wantErr: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			_, err := From(tt.input)
			if (err != nil) != tt.wantErr {
				t.Errorf("From() error = %v, wantErr %v", err, tt.wantErr)
			}
		})
	}
}

func TestToHex(t *testing.T) {
	addr, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	hex := addr.ToHex()

	if hex != "0x742d35cc6634c0532925a3b844bc9e7595f251e3" {
		t.Errorf("ToHex() = %s, want 0x742d35cc6634c0532925a3b844bc9e7595f251e3", hex)
	}
}

func TestToBytes(t *testing.T) {
	addr, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	b := addr.ToBytes()

	if len(b) != Size {
		t.Errorf("ToBytes() length = %d, want %d", len(b), Size)
	}
}

func TestEquals(t *testing.T) {
	addr1, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	addr2, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	addr3, _ := FromHex("0x0000000000000000000000000000000000000000")

	if !addr1.Equals(addr2) {
		t.Error("Equals() expected true for equal addresses")
	}

	if addr1.Equals(addr3) {
		t.Error("Equals() expected false for different addresses")
	}
}

func TestCompare(t *testing.T) {
	addr1, _ := FromHex("0x0000000000000000000000000000000000000001")
	addr2, _ := FromHex("0x0000000000000000000000000000000000000002")
	addr3, _ := FromHex("0x0000000000000000000000000000000000000001")

	if addr1.Compare(addr2) >= 0 {
		t.Error("Compare() expected negative for addr1 < addr2")
	}

	if addr2.Compare(addr1) <= 0 {
		t.Error("Compare() expected positive for addr2 > addr1")
	}

	if addr1.Compare(addr3) != 0 {
		t.Error("Compare() expected 0 for equal addresses")
	}
}

func TestLessThan(t *testing.T) {
	addr1, _ := FromHex("0x0000000000000000000000000000000000000001")
	addr2, _ := FromHex("0x0000000000000000000000000000000000000002")

	if !addr1.LessThan(addr2) {
		t.Error("LessThan() expected true")
	}

	if addr2.LessThan(addr1) {
		t.Error("LessThan() expected false")
	}
}

func TestGreaterThan(t *testing.T) {
	addr1, _ := FromHex("0x0000000000000000000000000000000000000001")
	addr2, _ := FromHex("0x0000000000000000000000000000000000000002")

	if addr1.GreaterThan(addr2) {
		t.Error("GreaterThan() expected false")
	}

	if !addr2.GreaterThan(addr1) {
		t.Error("GreaterThan() expected true")
	}
}

func TestIsZero(t *testing.T) {
	zeroAddr := Zero()
	nonZeroAddr, _ := FromHex("0x0000000000000000000000000000000000000001")

	if !zeroAddr.IsZero() {
		t.Error("IsZero() expected true for zero address")
	}

	if nonZeroAddr.IsZero() {
		t.Error("IsZero() expected false for non-zero address")
	}
}

func TestIsValid(t *testing.T) {
	tests := []struct {
		name  string
		input string
		want  bool
	}{
		{
			name:  "valid with 0x",
			input: "0x742d35cc6634c0532925a3b844bc9e7595f251e3",
			want:  true,
		},
		{
			name:  "valid without 0x",
			input: "742d35cc6634c0532925a3b844bc9e7595f251e3",
			want:  true,
		},
		{
			name:  "invalid length",
			input: "0x742d35cc",
			want:  false,
		},
		{
			name:  "invalid characters",
			input: "0x742d35cc6634c0532925a3b844bc9e7595f251eZ",
			want:  false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsValid(tt.input); got != tt.want {
				t.Errorf("IsValid() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestString(t *testing.T) {
	addr, _ := FromHex("0x742d35cc6634c0532925a3b844bc9e7595f251e3")
	s := addr.String()

	if s != "0x742d35cc6634c0532925a3b844bc9e7595f251e3" {
		t.Errorf("String() = %s, want 0x742d35cc6634c0532925a3b844bc9e7595f251e3", s)
	}
}

func TestZero(t *testing.T) {
	addr := Zero()

	if !addr.IsZero() {
		t.Error("Zero() should return zero address")
	}

	if addr.ToHex() != "0x0000000000000000000000000000000000000000" {
		t.Errorf("Zero() hex = %s, want 0x0000000000000000000000000000000000000000", addr.ToHex())
	}
}
